plugins/cuda/cuda_plugin.c

```diff
@@ -0,0 +1,406 @@
+		return execvp(args[0], (char **)args);
+	} else { // parent
+		close(fd[WRITE]);
+
+		int bytes_read = read(fd[READ], buf, buf_size);
+		if (bytes_read > 0) {
+			buf[bytes_read - 1] = '\0';
+		}
+
+		int status;
+		waitpid(child_pid, &status, 0);
+		if (!WIFEXITED(status)) {
+			close(fd[READ]);
+			return -1;
+		}
+
+		close(fd[READ]);
+
+		return WEXITSTATUS(status);
+	}
+}
+
+static bool cuda_checkpoint_supports_flag(const char *flag)
+{
+	char msg_buf[2048];
+	const char *args[] = { CUDA_CHECKPOINT, "-h", NULL };
+	int ret = launch_cuda_checkpoint(args, msg_buf, sizeof(msg_buf));
+	if (ret != 0) {
+		pr_err("Failed to launch cuda-checkpoint utility, check that the utility is present in your $PATH\n");
+		return false;
+	}
+
+	if (strstr(msg_buf, flag) == NULL) {
+		return false;
+	}
+
+	return true;
+}
+
+/* Retrieve the cuda restore thread TID from the root pid */
+static int get_cuda_restore_pid(int root_pid)
+{
+	char pid_buf[16];
+	char pid_out[CUDA_CKPT_BUF_SIZE];
+
+	snprintf(pid_buf, sizeof(pid_buf), "%d", root_pid);
+
+	const char *args[] = { CUDA_CHECKPOINT, "--get-restore-tid", "--pid", pid_buf, NULL };
+	int ret = launch_cuda_checkpoint(args, pid_out, sizeof(pid_out));
+	if (ret != 0) {
+		pr_err("Failed to launch cuda-checkpoint to retrieve restore tid: %s\n", pid_out);
+		return -1;
+	}
+
+	return atoi(pid_out);
+}
+
+static int cuda_process_checkpoint_action(int pid, const char *action, unsigned int timeout, char *msg_buf,
+					  int buf_size)
+{
+	char pid_buf[16];
+	char timeout_buf[16];
+
+	snprintf(pid_buf, sizeof(pid_buf), "%d", pid);
+
+	const char *args[] = { CUDA_CHECKPOINT, "--action", action, "--pid", pid_buf, NULL /* --timeout */,
+			       NULL /* timeout_val */, NULL };
+	if (timeout > 0) {
+		snprintf(timeout_buf, sizeof(timeout_buf), "%d", timeout);
+		args[5] = "--timeout";
+		args[6] = timeout_buf;
+	}
+
+	return launch_cuda_checkpoint(args, msg_buf, buf_size);
+}
+
+static int interrupt_restore_thread(int restore_pid)
+{
+	/* Since we resumed a thread that CRIU previously already froze we need to
+	 * INTERRUPT it once again, task was already SEIZE'd so we don't need to do
+	 * a compel_interrupt_task()
+	 */
+	if (ptrace(PTRACE_INTERRUPT, restore_pid, NULL, 0)) {
+		pr_err("Could not interrupt cuda restore tid %d after checkpoint, process may be in strange state\n",
+		       restore_pid);
+		return -1;
+	}
+
+	/* compel_wait_task() will fail here as expected but the work we need it to
+	 * do is actually done before the sigstop check that fails. Without this
+	 * compel_stop_on_syscall() will hang waiting for the task to hit a
+	 * sigreturn when cleaning up the parasite
+	 */
+	struct proc_status_creds creds;
+	if (compel_wait_task(restore_pid, -1, parse_pid_status, NULL, &creds.s, NULL)) {
+		pr_debug("wait_task failed as expected. ignoring\n");
+	}
+
+	return 0;
+}
+
+int cuda_plugin_checkpoint_devices(int pid)
+{
+	int restore_pid;
+	char msg_buf[CUDA_CKPT_BUF_SIZE];
+	int ret = 0;
```

Comment:
nit: I think it is better to initialize it to -1 and set it to 0 after all actions. In this case, you will not forget to set it to -1 on all error path.  We usually call it `exit_code`, because `ret` is often used for intermediate calls.